src/test/java/com/squareup/javapoet/TypeNameTest.java

```diff
@@ -41,6 +72,31 @@
     TypeName.get(recursiveEnum.getGenericParameterTypes()[0]);
   }
 
+  @Test
```

Comment:
FYI this new tests fail before this change, i.e. the ParameterizedTypeName returned is a raw type of TestGeneric.Inner instead of a TestGeneric<String>Inner.

The other 2 additional tests are to make sure that the newly added code doesn't break old behavior.